### PR TITLE
CSRF token cookie -> header 

### PR DESCRIPTION
**Pull request by reasonerjt**
The current approach will prevent the effectiveness of `Cache-Control`
header and gorilla's library add `Vary:Cookie` header in all responses.

This PR sets the token in a header of response so the response can be
cached when needed.  It involves code change in both UI and backend middleware.

### CODE REVIEW

File: src/portal/src/app/services/intercept-http.service.ts [hunk 1]
```diff
@@ -1,28 +1,56 @@
...
-          request = request.clone({ headers: request.headers.set('X-Harbor-CSRF-Token', Xsrftoken) });
-          return next.handle(request);
-        }
+    // Get the csrf token from localstorage
+    const token = localStorage.getItem("__csrf");
+    if (token) {
+      // Clone the request and replace the original headers with
+      // cloned headers, updated with the csrf token.
+      // not for 'get' request
+      if (request.method && request.method.toLocaleLowerCase() !== 'get') {
```
Comment (reasonerjt) [inline src/portal/src/app/services/intercept-http.service.ts:20]: The safe methods should  include:
"GET", "HEAD", "OPTIONS", "TRACE"

**Review by ywk253100 — APPROVED** [review]

File: src/portal/src/app/services/intercept-http.service.ts (hunk 1 above)
Comment (AllForNothing) [reply src/portal/src/app/services/intercept-http.service.ts]: fixed

File: src/portal/src/app/services/intercept-http.service.ts [hunk 2]
```diff
@@ -1,28 +1,58 @@
...
+           }
+         }
+       }))
+      .pipe(
+       catchError(error => {
+         if (error.status === 403) {
+           const csrfToken = localStorage.getItem("__csrf");
+           if (csrfToken) {
+             request = request.clone({ headers: request.headers.set('X-Harbor-CSRF-Token', csrfToken)});
+             return next.handle(request);
```
Comment (jwangyangls) [inline src/portal/src/app/services/intercept-http.service.ts:51]: If there is one who doesn't know ，the error results in a 403 all the time, and then there will always be requests

Comment (AllForNothing) [reply src/portal/src/app/services/intercept-http.service.ts:51]: This is the same as before。 if csrf token is timeout, back end will  return 403 error and  a valid csrf token,  so the next request will be successful

**Review by jwangyangls — APPROVED** [review]

**Pull request merged**